BuildScript/BuildTasks/MetadataValueDictionary.cs

```diff
@@ -67,5 +67,22 @@ public bool ContainsKey (string key)
     {
       return _rawDictionary.ContainsKey (key);
     }
+
+    public IEnumerable<KeyValuePair<string, string>> GetAllMatches (IEnumerable<string> possibleKeys, Func<Exception> createEmptySequenceException)
+    {
+      var enumeratedPossibleKeys = possibleKeys.ToArray();
+
+      var intersection = _rawDictionary.Keys.Intersect (enumeratedPossibleKeys, _dictionaryStringComparer).ToArray();
+
+      if (intersection.Length == 0)
```

Comment:
GetAllMatches doens't imply there's an exception possible for no matches. From the usage, there's also no need to unify this into a single operation. => please drop exception logic. In case: Single works differently and has the exception semantics included by the .net framework. Plus, you could not distinguish NULL and NOT FOUND with Single without an exception.